perf/perf_core_imc_event.py

```diff
@@ -0,0 +1,111 @@
+            if dist.name in ['Ubuntu']:
+                pkgs.extend(['linux-tools-common'])
+        elif dist.name in ['centos', 'fedora', 'rhel', 'SuSE']:
+            pkgs = ['perf']
+        else:
+            self.cancel("perf is not supported on %s" % dist.name)
+
+        for pkg in pkgs:
+            if not smg.check_installed(pkg) and not smg.install(pkg):
+                self.cancel(
+                    "Package %s is missing/could not be installed" % pkg)
+        process.run("ppc64_cpu --frequency -t 10 &", shell=True,
+                    ignore_status=True, verbose=True, ignore_bg_processes=True)
+
+        # collect all imc events
+        self.list_core_imc = []
+        self.list_thread_imc = []
+        self.list_trace_imc = []
+        for line in process.get_command_output_matching('perf list', 'imc'):
+            line = "%s" % line.split()[0]
+            if 'core_imc' in line:
+                self.list_core_imc.append(line)
+            elif 'thread_imc' in line:
+                self.list_thread_imc.append(line)
+            elif 'trace_imc' in line:
+                self.list_trace_imc.append(line)
+
+        dmesg.clear_dmesg()
+
+    def parse_op(self, cmd):
+        fail_count = 0
+        result = process.run(cmd, shell=True, sudo=True)
+        output = result.stdout.decode() + result.stderr.decode()
+        if ("not counted" in output) or ("not supported" in output):
+            fail_count = fail_count + 1
+        if fail_count > 0:
+            self.fail("%s : command failed" % cmd)
+
+    def imc_events(self, event):
+        for line in event:
+            if line in self.list_core_imc:
+                cmd = "perf stat -e %s -I 1000 sleep 5" % line
+                self.parse_op(cmd)
+            if line in self.list_thread_imc:
+                cmd = "perf stat -e %s -I 1000 ls -R /usr/ > /dev/null" % line
+                self.parse_op(cmd)
+            if line in self.list_trace_imc:
+                cmd = "perf record -o perf.data -e %s -C 0 sleep 5" % line
+                process.run(cmd, shell=True, sudo=True)
+                res = process.run(
+                    "perf report --stdio -i perf.data", shell=True, sudo=True)
+                if "data has no samples!" in res.stderr.decode():
+                    self.fail("trace_imc perf.data sample not captured")
+
+    def test_core_imc(self):
+        for event in [self.list_core_imc, self.list_thread_imc,
+                      self.list_trace_imc]:
+            self.imc_events(event)
+
+        # negative testcase running two different imc events parallely
+        process.run("perf stat -e core_imc/CPM_CCYC/ -I 1000 &",
+                    sudo=True, shell=True, ignore_bg_processes=True)
+        if not process.system("perf stat -e thread_imc/CPM_CCYC/ -I 1000 ls",
+                              ignore_status=True, sudo=True):
+            self.fail("expected a failure, able to run different imc events "
+                      "parallely")
```

Comment:
failure message confusing, if the even run fails,
you said `"expected a failure,  able to run different imc events   "parallely")`

can you elaborate little bit to get exact glimpse of it?

Reply:
@Naresh-ibm if we run two different imc events at the same time it will give an error. for example:
here i am running core_imc event in background and trying to run thread_imc event in parallel, it gives error with non zero exit status.

```
[root@ ~]# perf stat -e core_imc/CPM_CCYC/ -I 1000 &
[1] 37976

[root@ ~]# perf stat -e thread_imc/CPM_CCYC/ -I 1000 ls
Error:
The sys_perf_event_open() syscall returned with 16 (Device or resource busy) for event (thread_imc/CPM_CCYC/).
/bin/dmesg | grep -i perf may provide additional information.
```

Reply:
@disgoel is this the expected failure?
if it is expected failure why we fail the test here. we can just use self.log.info() to print the message, as that is a pass case for us.

Reply:
@Naresh-ibm we are expecting a failure here, if exit status is non-zero means command failed and test will pass.
In below condition we are checking if exit status is zero then test should fail.

```
        # negative testcase running two different imc events parallely
        process.run("perf stat -e core_imc/CPM_CCYC/ -I 1000 &",
                    sudo=True, shell=True, ignore_bg_processes=True)
        if not process.system("perf stat -e thread_imc/CPM_CCYC/ -I 1000 ls",
                              ignore_status=True, sudo=True):
            self.fail("test failed because able to run two different imc "
                      "events parallely")
```

Reply:
ok, got it. the if check fails if it is 0 and passes if it is non-zero and the return code is wise versa.